File: Blind 75/01 - Arrays and Hashing/242-valid-anagram.py
'''
Given two strings s and t, return true if t is an anagram of s, and false otherwise.

An Anagram is a word or phrase formed by rearranging the letters of a different word or phrase, typically using all the original letters exactly once.
'''
from time import time


class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False

        s_table, t_table = {}, {}
        for i in range(len(s)):
            s_table[s[i]] = s_table.get(s[i], 0) + 1
            t_table[t[i]] = t_table.get(t[i], 0) + 1

        if s_table == t_table:
            return True
        else:
            return False

    def reference(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False

        countS, countT = {}, {}

        for i in range(len(s)):
            countS[s[i]] = 1 + countS.get(s[i], 0)
            countT[t[i]] = 1 + countT.get(t[i], 0)
        for c in countS:
            if countS[c] != countT.get(c, 0):
                return False
        return True

    def quantify(self, test_cases, runs=50000):
        sol_start = time()
        for i in range(runs):
            for case in test_cases:
                if i == 0:
                    print(self.isAnagram(*case))
                else:
                    self.isAnagram(*case)
        print(f'Runtime for our solution: {time() - sol_start}\n')

        ref_start = time()
        for i in range(0, runs):
            for case in test_cases:
                if i == 0:
                    print(self.reference(*case))
                else:
                    self.reference(*case)
        print(f'Runtime for reference: {time() - ref_start}')


if __name__ == '__main__':
    test = Solution()
    test_cases = [('anagram', 'nagaram'), ('rat', 'car')]
    test.quantify(test_cases)

File: Blind 75/01 - Arrays and Hashing/JavaScript/242-valid-anagram.js
const { performance } = require("perf_hooks");

class Solution {
  method(s, t) {
    if (s.length !== t.length) return false;

    let charCount = {};

    for (let i = 0; i < s.length; i++) {
      charCount[s[i]] = (charCount[s[i]] || 0) + 1;
      charCount[t[i]] = (charCount[t[i]] || 0) - 1;
    }

    for (const k in charCount) {
      if (charCount[k] !== 0) return false;
    }

    return true;
  }

  reference(s, t) {
    if (s.length !== t.length) {
      return false;
    }
    let sMap = {};
    let tMap = {};
    for (let i = 0; i < s.length; i++) {
      if (sMap.hasOwnProperty(s[i])) {
        sMap[s[i]]++;
      } else {
        sMap[s[i]] = 1;
      }
      if (tMap.hasOwnProperty(t[i])) {
        tMap[t[i]]++;
      } else {
        tMap[t[i]] = 1;
      }
    }
    for (let k in sMap) {
      if (sMap[k] !== tMap[k]) {
        return false;
      }
    }
    return true;
  }

  quantify(testCases, runs = 1e6) {
    const runsArr = Array.from({ length: runs });
    const solStart = performance.now();
    runsArr.map((_, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.method(...input));
        else this.method(...input);
      });
    });
    console.log(
      `Runtime for solution: ${(performance.now() - solStart) / 1000}\n`
    );

    const refStart = performance.now();
    runsArr.map((_, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.reference(...input));
        else this.reference(...input);
      });
    });
    console.log(
      `Runtime for reference: ${(performance.now() - refStart) / 1000}`
    );
  }
}

const test = new Solution();
const testCases = [
  ["anagram", "nagaram"],
  ["rat", "car"],
];
test.quantify(testCases);
